rust/connlib/tunnel/src/client.rs

```diff
@@ -281,37 +281,61 @@ fn send_dns_answer(
+///
+/// Internally, this composes a [`snownet::ClientNode`] with firezone's policy engine around resources.
+/// Clients differ from gateways in that they also implement a DNS resolver for DNS resources.
+/// They also initiate connections to gateways based on packets sent to resources.
 pub struct ClientState {
+    /// The [`snownet::ClientNode`].
+    ///
+    /// Manages wireguard tunnels to gateways.
+    node: ClientNode<GatewayId, RelayId>,
+    /// All peers gateways we are connected to and the associated, connection-specific state.
+    peers: PeerStore<GatewayId, GatewayOnClient>,
+    /// Which resources are trying to connect to.
     awaiting_connection: HashMap<ResourceId, AwaitingConnectionDetails>,
+
+    /// Tracks, which gateway to use for a particular resource.
     resources_gateways: HashMap<ResourceId, GatewayId>,
+    /// The site a gateway belongs to.
+    gateways_site: HashMap<GatewayId, SiteId>,
+    /// The online/offline status of a site.
+    sites_status: HashMap<SiteId, Status>,
 
-    pub dns_resources_internal_ips: HashMap<DnsResource, HashSet<IpAddr>>,
+    /// All DNS resources we know about, indexed by their domain (could be wildcard domain like *.mycompany.com).
     dns_resources: HashMap<String, ResourceDescriptionDns>,
+    /// All CIDR resources we know about, indexed by the IP range they cover (like `1.1.0.0/8`).
     cidr_resources: IpNetworkTable<ResourceDescriptionCidr>,
-    pub resource_ids: HashMap<ResourceId, ResourceDescription>,
-    pub deferred_dns_queries: HashMap<(DnsResource, Rtype), IpPacket<'static>>,
-
-    peers: PeerStore<GatewayId, GatewayOnClient>,
-
-    node: ClientNode<GatewayId, RelayId>,
-
-    pub ip_provider: IpProvider,
-
-    dns_mapping: BiMap<IpAddr, DnsServer>,
+    /// All resources indexed by their ID.
+    resource_ids: HashMap<ResourceId, ResourceDescription>,
 
-    buffered_events: VecDeque<ClientEvent>,
-    interface_config: Option<InterfaceConfig>,
-    buffered_packets: VecDeque<IpPacket<'static>>,
+    /// The DNS resolvers configured on the system prior to connlib starting up.
+    system_resolvers: Vec<IpAddr>,
 
     /// DNS queries that we need to forward to the system resolver.
     buffered_dns_queries: VecDeque<DnsQuery<'static>>,
 
+    /// The (internal) IPs we have assigned for a certain DNS resource.
+    ///
+    /// We assign one internal IP per externally resolved IP.
+    dns_resources_internal_ips: HashMap<DnsResource, HashSet<IpAddr>>,
+    /// DNS queries we can only answer once we have connected to the resource.
+    ///
+    /// See [`dns::ResolveStrategy`] for details.
+    deferred_dns_queries: HashMap<(DnsResource, Rtype), IpPacket<'static>>,
+    /// Hands out IPs for DNS resources.
+    ip_provider: IpProvider,
+    /// Maps from connlib-assigned IP of a DNS server back to the originally configured system DNS resolver.
+    dns_mapping: BiMap<IpAddr, DnsServer>,
+    /// When to next refresh DNS resources.
```

Comment:
What does this do? Do we redo DNS resolving even while connected to a Resource? Or is this the DNS TTL?

Reply:
Good question, let me dig into this.

Reply:
It is the former. I clarified this in a comment.

Reply:
the only reason we don't use the TTL here is because we don't have it, we would need to use something like hickory on the gateway too to obtain it.

with the DNS refactor we are doing now this will go away anyways and refreshing will look really different with a whole new set of problems :stuck_out_tongue: